pytorch_lightning/core/datamodule.py

```diff
@@ -381,8 +433,13 @@ def wrapped_fn(*args: str, **kwargs: Optional[str]) -> Any:
                 has_run = obj._has_prepared_data
                 obj._has_prepared_data = True
 
-            if not has_run:
-                return fn(*args, **kwargs)
+            if has_run:
+                rank_zero_warn(
+                    f"DataModule.{name} has already been called, so it will not be called again. "
+                    f"In v1.6 this behavior will change to always call DataModule.{name}."
+                )
```

Comment:
You can take this test, move it to deprecated_api, and use it to test the warning:

https://github.com/PyTorchLightning/pytorch-lightning/blob/906c067b07a7caa148615880792e65bb779ee3c0/tests/core/test_datamodules.py#L529-L569